cabal-install/Distribution/Client/Setup.hs

```diff
@@ -387,18 +399,96 @@ instance Monoid GlobalFlags where
     where combine field = field a `mappend` field b
 
 withGlobalRepos :: Verbosity -> GlobalFlags -> ([Repo] -> IO a) -> IO a
-withGlobalRepos _verbosity globalFlags callback =
-    callback $ remoteRepos ++ localRepos
+withGlobalRepos verbosity globalFlags callback =
+    withSecureRepos verbosity secureRepos $ \secureRepos' ->
+      callback $ secureRepos' ++ remoteRepos ++ localRepos
   where
+    secureRepos =
+      [ (remote, cacheDir)
+      | remote <- fromNubList $ globalRemoteRepos globalFlags
+      , remoteRepoSecure remote
+      , let cacheDir = fromFlag (globalCacheDir globalFlags)
+                   </> remoteRepoName remote ]
     remoteRepos =
       [ RepoRemote remote cacheDir
       | remote <- fromNubList $ globalRemoteRepos globalFlags
+      , not (remoteRepoSecure remote)
       , let cacheDir = fromFlag (globalCacheDir globalFlags)
                    </> remoteRepoName remote ]
     localRepos =
       [ RepoLocal local
       | local <- fromNubList $ globalLocalRepos globalFlags ]
 
+-- | Initialize the provided secure repositories
+--
+-- Assumed invariant: `remoteRepoSecure` should be set for all these repos.
+withSecureRepos :: forall a.
+                   Verbosity
+                -> [(RemoteRepo, FilePath)]  -- ^ Repos and cacheDirs
+                -> ([Repo] -> IO a)          -- ^ Callback
+                -> IO a
+withSecureRepos verbosity repos callback = go [] repos
+  where
+    go :: [Repo] -> [(RemoteRepo, FilePath)] -> IO a
+    go acc []                = callback (reverse acc)
+    go acc ((r,cacheDir):rs) = do
+      cachePath <- Sec.makeAbsolute $ Sec.fromFilePath cacheDir
+      withSecureRepo verbosity r cachePath $ \r' ->
+        go (RepoSecure r cacheDir r':acc) rs
+
+-- | Initialize the given secure repo
+--
+-- The security library has its own concept of a "local" repository, distinct
```

Comment:
Hmmm, this I am not so keen on. Cabal's concept of a local repository is quite different from any other kind of repository (different format, for one). This would also require a further change to the config file.